Implement Auto Craft Ingredients option

MassCraxx opened the pull request:
This should add an new TickBox option in the UI tab to let the AutoCook consider ingredients from closed cans in the vicinity.
Closed cans chosen for the recipe and all required tools will be taken into the inventory, used to craft the usable opened ingredient and then finally used in the AutoCook process.

Open:
- <s>Recipe code is unstable and will get a rework from the AutoEat mod</s>
- <s>Context menu option will not show up if there are only valid closed cans around</s>
- <s>UI</s>
- <s>optimizations, caching, reduce load</s>

Known Issues:
- if there is only a craftable valid spice in the vicinity (e.g Gravy Mix in Roasting Pan), autocook will be available, but simply craft the spice and stop doing anything, as the recipe can not be started with a spice... Looking for a way to filter it out pre-craft without creating a fake item..
-- turns out gravy is the only vanilla case where this happens. Excluding this one seems good enough

- <s>choosing algorithm prefers crafting items, resulting in more crafts than uses if available</s>
- <s>as long as there are possible craftable ingredients around, autocook will not stop until its all crafted</s>

MassCraxx committed:
implement valid closed cans lookup for selected recipe,
implement auto open cans, including tool handling,
recipe handling unstable, context menu needs rework

katupia commented on AutoCook/media/lua/client/AutoCook.lua:
```diff
@@ -396,3 +495,6 @@ function AutoCook:new(playerObj, recipe, baseItem)
 end
+
+-- load persisted values on reload script
+AutoCook:init(getPlayer())
```
Keeping active debug code is only ok within a "false" condition by default. Maybe add some AutoCook.Debug same as there is an AutoCook.Verbose

MassCraxx replied:
agreed, will only happen in debug mode now

MassCraxx committed:
rework OpenCans to AutoCraftIngredients, now considering all food crafts,
precache relevant recipes on init,
rework context menu to always show AutoCook,
include tooltip in context menu entry,
generalize modData handling,
generalize UI element creation,
add AutoCraftIngredients UI TickBox

MassCraxx commented:
Reworked this feature to work with all crafted food ingredients, not just cans.
Also reworked ContextMenu for it to show up under new conditions. Has tooltip now:
![image](https://github.com/user-attachments/assets/bdf5b994-0580-4a18-9c4d-595c4daefb46)
![image](https://github.com/user-attachments/assets/94a62f56-7da5-4fed-90d6-c4b7386ce300)

added UI for AutoCraft
![image](https://github.com/user-attachments/assets/457d8e36-c80d-4298-af79-d64aa58c8ac7)

MassCraxx committed:
use result item icon in context tooltip

MassCraxx commented:
Changed icon to show result.
![image](https://github.com/user-attachments/assets/ac634198-14ed-48bb-9c33-578c9be9fb95)

MassCraxx committed:
cleanup

MassCraxx committed:
fix start evo recipe with only gravy

MassCraxx committed:
ignore possible crafted ingredients that are available already,
cache created potential craft result items for comparison

MassCraxx committed:
fix autocook not stopping to craft on finished meal

MassCraxx committed:
fix autocook now stopping too early..

katupia committed:
Integration layer for MassCraxx propositions 2

AutoCook File split
tooltip typo
variable scope
x4 Font handling
